mmv1/third_party/terraform/functions/name_from_id_internal_test.go

```diff
@@ -0,0 +1,97 @@
+package functions
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/function"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+)
+
+func TestFunctionRun_name_from_id(t *testing.T) {
+	t.Parallel()
+
+	name := "foobar"
+
+	// Happy path inputs
+	validId := fmt.Sprintf("projects/my-project/zones/us-central1-c/instances/%s", name)
+	validSelfLink := fmt.Sprintf("https://www.googleapis.com/compute/v1/%s", validId)
+	validOpStyleResourceName := fmt.Sprintf("//gkehub.googleapis.com/projects/my-project/locations/us-central1/memberships/%s", name)
+
+	// Unhappy path inputs
+	invalidInput := "this isn't a URI or id"
+
+	testCases := map[string]struct {
+		request  function.RunRequest
+		expected function.RunResponse
+	}{
+		"it returns the expected output value when given a valid resource id input": {
+			request: function.RunRequest{
+				Arguments: function.NewArgumentsData([]attr.Value{types.StringValue(validId)}),
+			},
+			expected: function.RunResponse{
+				Result: function.NewResultData(types.StringValue(name)),
+			},
+		},
+		"it returns the expected output value when given a valid resource self_link input": {
+			request: function.RunRequest{
+				Arguments: function.NewArgumentsData([]attr.Value{types.StringValue(validSelfLink)}),
+			},
+			expected: function.RunResponse{
+				Result: function.NewResultData(types.StringValue(name)),
+			},
+		},
+		"it returns the expected output value when given a valid OP style resource name input": {
+			request: function.RunRequest{
+				Arguments: function.NewArgumentsData([]attr.Value{types.StringValue(validOpStyleResourceName)}),
+			},
+			expected: function.RunResponse{
+				Result: function.NewResultData(types.StringValue(name)),
+			},
+		},
+		"it returns an error when given input with no submatches": {
```

Comment:
There isn't a test for multiple submatches because you cannot have multiple ends of a string (`$` is used in the regex)

Reply:
Makes sense!